### PR TITLE
Fix a class of logcontext leaks

### PR DESCRIPTION
**Pull request by richvdh**
So, it turns out that if you have a first `Deferred` `D1`, you can add a callback which returns another `Deferred` `D2`, and `D2` must then complete before any further callbacks on `D1` will execute (and later callbacks on `D1` get the *result* of `D2` rather than `D2` itself).

So, `D1` might have `called=True` (as in, it has started running its callbacks), but any new callbacks added to `D1` won't get run until `D2` completes - so if you `yield D1` in an `inlineCallbacks` function, your `yield` will 'block'.

In conclusion: some of our assumptions in `logcontext` were invalid. We need to make sure that we don't optimise out the logcontext juggling when this situation happens. Fortunately, it is easy to detect by checking `D1.paused`.

### CODE REVIEW

File: tests/util/test_logcontext.py [hunk 1]
```diff
@@ -80,20 +84,31 @@ def check_logcontext():
...
+    def test_run_in_background_with_non_blocking_fn(self):
         @defer.inlineCallbacks
         def nonblocking_function():
             with logcontext.PreserveLoggingContext():
                 yield defer.succeed(None)
 
-        return self._test_preserve_fn(nonblocking_function)
+        return self._test_run_in_background(nonblocking_function)
+
+    @unittest.DEBUG
```
Comment (erikjohnston) [inline tests/util/test_logcontext.py]: What's this?

Comment (richvdh) [reply tests/util/test_logcontext.py]: it's a fail

**Comment by richvdh** [comment]
removed the spurious `@DEBUG`. #3172 has the fixes to the tests.
